lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp

```diff
@@ -616,4 +624,133 @@ int amf_smf_handle_ip_address_response(
+
+  DLNASTransportMsg& dlmsg =
+      msg.security_protected.plain.amf.msg.downlinknas5gtransport;
+
+  // AmfHeader
+  dlmsg.extended_protocol_discriminator.extended_proto_discriminator =
+      M5G_MOBILITY_MANAGEMENT_MESSAGES;
+  len++;
+  dlmsg.spare_half_octet.spare  = 0x00;
+  dlmsg.sec_header_type.sec_hdr = SECURITY_HEADER_TYPE_NOT_PROTECTED;
+  len++;
+  dlmsg.message_type.msg_type = DLNASTRANSPORT;
+  len++;
+  dlmsg.payload_container.iei = PAYLOAD_CONTAINER;
+
+  // SmfMsg
+  dlmsg.payload_container_type.iei      = 0;
+  dlmsg.payload_container_type.type_val = N1_SM_INFO;
+  len++;
+  dlmsg.pdu_session_identity.iei = PDU_SESSION_IDENTITY;
+  len++;
+  dlmsg.pdu_session_identity.pdu_session_id = session_id;
+  len++;
+
+  SmfMsg& pdu_sess_est_reject = dlmsg.payload_container.smf_msg;
+  // header
+  pdu_sess_est_reject.header.extended_protocol_discriminator =
+      M5G_SESSION_MANAGEMENT_MESSAGES;
+  pdu_sess_est_reject.header.pdu_session_id = session_id;
+  pdu_sess_est_reject.header.message_type   = PDU_SESSION_ESTABLISHMENT_REJECT;
+  pdu_sess_est_reject.header.procedure_transaction_id = pti;
+  // container_len++;
+
+  // Smf NAS message
+  pdu_sess_est_reject.msg.pdu_session_estab_reject
+      .extended_protocol_discriminator.extended_proto_discriminator =
+      M5G_SESSION_MANAGEMENT_MESSAGES;
+  container_len++;
+  pdu_sess_est_reject.msg.pdu_session_estab_reject.pdu_session_identity
+      .pdu_session_id = session_id;
+  container_len++;
+  pdu_sess_est_reject.msg.pdu_session_estab_reject.pti.pti = pti;
+  container_len++;
+  pdu_sess_est_reject.msg.pdu_session_estab_reject.message_type.msg_type =
+      PDU_SESSION_ESTABLISHMENT_REJECT;
+  container_len++;
+  pdu_sess_est_reject.msg.pdu_session_estab_reject.m5gsm_cause.cause_value =
+      cause;
+  container_len++;
+
+  dlmsg.payload_container.len = container_len;
+  len += 2;  // 2 bytes for container.len
+  len += dlmsg.payload_container.len;
+
+  /* Ciphering algorithms, EEA1 and EEA2 expects length to be mode of 4,
+   * so length is modified such that it will be mode of 4
+   */
+  AMF_GET_BYTE_ALIGNED_LENGTH(len);
+  if (msg.header.security_header_type != SECURITY_HEADER_TYPE_NOT_PROTECTED) {
+    amf_msg_header* header = &msg.security_protected.plain.amf.header;
+    /*
+     * Expand size of protected NAS message
+     */
+    len += NAS_MESSAGE_SECURITY_HEADER_SIZE;
+    /*
+     * Set header of plain NAS message
+     */
+    header->extended_protocol_discriminator = M5G_MOBILITY_MANAGEMENT_MESSAGES;
+    header->security_header_type = SECURITY_HEADER_TYPE_NOT_PROTECTED;
+  }
+
+  buffer = bfromcstralloc(len, "\0");
+  bytes  = nas5g_message_encode(
+      buffer->data, &msg, len, &ue_context->amf_context._security);
+  if (bytes > 0) {
+    buffer->slen = bytes;
+    amf_app_handle_nas_dl_req(ue_id, buffer, rc);
+
+  } else {
+    OAILOG_WARNING(LOG_AMF_APP, "NAS encode failed \n");
+    bdestroy_wrapper(&buffer);
```

Comment:
Woulnt we want to clean up in the success case? after the amf_app_handle_nas_dl_req?